text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+
+- **Shipyard:** A shipyard is the declarative means to divide an environment into project stages and to specify task sequences for each stage.
+
+- **Sequence:** A sequence declares a set of tasks, which are triggered by an external event (aka. domain event, e.g.: artifact built → deploy artifact, a problem occurred → execute remediation, etc.). A sequence executes the tasks in the given order. 
+
+- **Task:** A task is the smallest executable unit of a sequence. A task is kicked-off by a *triggered* event and has a life cycle of a *start* and *finished* event. 
+
+*Current situation and Problem(s):* The current shipyard specification allows dividing an environment into project stages. In each stage, two ways for deploying (direct, blue/green) and testing (functional/performance) an artifact are provided. Besides, an SRE can specify whether a project stage is capable of handling problems by remediation actions. This specification has some limitations: 
+
+- Certain tasks are defined implicitly without giving an SRE the possibility to remove/move the task, e.g.:
+
+  - The evaluation always happens after a test execution even though the SRE already knows that no meaningful metric values are available (e.g., functional tests don't provide performance data). 
+
+  - The promotion of a deployed/released artifact happens automatically with no possibility to intervene.
+
+- It is not possible to add a task to a sequence, e.g., a task that takes care of pulling approval from a person in charge. 
+
+- The trigger (domain event) of a task sequence can not be set. 
+
+*Solution:* A revision of the shipyard specification v0.1.2 addresses the above-mentioned problems and gives SREs more flexibility in declaring processes. 
+
+## Internal details
+
+### Specification 
+
+The KEP proposes a new version for the shipyard specification as explained below. To get started, the following example of a shipyard with one stage is provided: 
+
+```yaml
+---
+version: 0.2.0
+kind: Shipyard
+metadata:
+  name: shipyard-abc
+spec:
+  stages:
+  - name: hardening
+    sequences:
+    - name: artifact-delivery
+      triggers:   
+      - dev.artifact-delivery.finished
+      tasks:
+      - name: deployment
+        properties:
+          strategy: blue_green
+          traffic: 100
+      - name: test
+        properties:
+          kind: functional
+      - name: evaluation 
+      - name: test
+        properties:
+          kind: performance
+      - name: evaluation
+      - name: release 
+```
+
+*Meta-data:*
+* **version**: The version of the shipyard specification. 
+* **kind**: Is set to `Shipyard`.
+* **metadata:** Contains at least the property *name*, which declares a unique name for the shipyard.
```

Comment:
How is the name unique? Are we checking that in our implementation? Does it always have to be the project name?

Reply:
I checked out how Kubernetes is defining this field. I found: https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields

The definition says: `Data that helps uniquely identify the object, including a name string, UID, and optional namespace` 

I would also go with this definition and would not check it in our implementation.